src/main/java/org/truffleruby/builtins/AmbiguousOptionalArgumentChecker.java

```diff
@@ -36,54 +40,83 @@ public static void verifyNoAmbiguousOptionalArguments(NodeFactory<? extends Ruby
-            }
-
-            Class<?> node = nodeFactory.getNodeClass();
+            final int opt = methodAnnotation.optional();
 
-            for (int i = 1; i <= opt; i++) {
-                boolean unguardedObjectArgument = false;
+            final Class<?> node = nodeFactory.getNodeClass();
+            for (Method method : specializations(node)) {
+                boolean unguardedParams = false;
                 StringBuilder errors = new StringBuilder();
-                for (Method method : node.getDeclaredMethods()) {
-                    if (method.isAnnotationPresent(Specialization.class)) {
-                        // count from the end to ignore optional VirtualFrame in front.
-                        Class<?>[] parameterTypes = method.getParameterTypes();
-                        int n = parameterTypes.length - i;
-                        if (methodAnnotation.rest()) {
-                            n--; // ignore final Object[] argument
-                        }
-                        Class<?> parameterType = parameterTypes[n];
-                        Parameter[] parameters = method.getParameters();
 
-                        Parameter parameter = parameters[n];
-                        boolean isNamePresent = parameter.isNamePresent();
-                        if (!isNamePresent) {
-                            AVAILABLE = SUCCESS = false;
-                            Log.LOGGER.warning("method parameters names are not available for " + method);
-                            return;
-                        }
-                        String name = parameter.getName();
+                Class<?>[] parameterTypes = method.getParameterTypes();
+                Parameter[] parameters = method.getParameters();
 
-                        if (parameterType == Object.class && !name.startsWith("unused") && !name.startsWith("maybe")) {
-                            String[] guards = method.getAnnotation(Specialization.class).guards();
-                            if (!isGuarded(name, guards)) {
-                                unguardedObjectArgument = true;
-                                errors.append("\"").append(name).append("\" in ").append(methodToString(method, parameterTypes, parameters)).append("\n");
-                            }
-                        }
+                int n = parameterTypes.length - 1;
+                // Ignore all the @Cached methods from our consideration.
+                while (n >= 0 && parameters[n].isAnnotationPresent(Cached.class)) {
+                    n--;
+                }
+                if (methodAnnotation.needsBlock()) {
+                    if (n < 0) {
+                        SUCCESS = false;
+                        Log.LOGGER.warning("invalid block method parameter position for " + method);
+                        continue;
                     }
+                    unguardedParams = unguardedParams | isParameterUnguarded(method, parameters, parameterTypes, n, errors);
```

Comment:
Should this be `||`?

Reply:
Ah, I see, it's to evaluate it no matter the value of `unguardedParams`.
Can `|=` be used as it's slightly shorter? (it feels clearer to me)

Reply:
Could you change to `|=`?